# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.cpp
# Commit: 25adf505426d
# Full Hash: 25adf505426d21f67e15d146f879b0ea53d343a7
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2022-03-23 04:09:13
# Regressor Bug: 1642412
# File Overlap Count: 19
# Description:
#   Bug 1642412 - wasm: Rollup of inline table.get/set for anyref. r=lth
#   
#   Rollup of the previous inline table.get/set patches.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D141024
# ==============================================================================

diff -r 43f7556db2e1 -r 25adf505426d js/src/wasm/WasmInstance.cpp
--- a/js/src/wasm/WasmInstance.cpp	Tue Mar 22 19:44:13 2022 +0000
+++ b/js/src/wasm/WasmInstance.cpp	Tue Mar 22 19:59:05 2022 +0000
@@ -998,9 +998,9 @@
   return 0;
 }
 
-/* static */ void* Instance::tableGet(Instance* instance, uint32_t index,
-                                      uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableGet.failureMode == FailureMode::FailOnInvalidRef);
+/* static */ void* Instance::tableGetFunc(Instance* instance, uint32_t index,
+                                          uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableGetFunc.failureMode == FailureMode::FailOnInvalidRef);
 
   JSContext* cx = instance->cx();
   const Table& table = *instance->tables()[tableIndex];
@@ -1009,19 +1009,13 @@
     return AnyRef::invalid().forCompiledCode();
   }
 
-  switch (table.repr()) {
-    case TableRepr::Ref:
-      return table.getAnyRef(index).forCompiledCode();
-    case TableRepr::Func: {
-      MOZ_RELEASE_ASSERT(!table.isAsmJS());
-      RootedFunction fun(cx);
-      if (!table.getFuncRef(cx, index, &fun)) {
-        return AnyRef::invalid().forCompiledCode();
-      }
-      return FuncRef::fromJSFunction(fun).forCompiledCode();
-    }
+  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
+  MOZ_RELEASE_ASSERT(!table.isAsmJS());
+  RootedFunction fun(cx);
+  if (!table.getFuncRef(cx, index, &fun)) {
+    return AnyRef::invalid().forCompiledCode();
   }
-  MOZ_CRASH("Should not happen");
+  return FuncRef::fromJSFunction(fun).forCompiledCode();
 }
 
 /* static */ uint32_t Instance::tableGrow(Instance* instance, void* initValue,
@@ -1050,9 +1044,9 @@
   return oldSize;
 }
 
-/* static */ int32_t Instance::tableSet(Instance* instance, uint32_t index,
-                                        void* value, uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableSet.failureMode == FailureMode::FailOnNegI32);
+/* static */ int32_t Instance::tableSetFunc(Instance* instance, uint32_t index,
+                                            void* value, uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableSetFunc.failureMode == FailureMode::FailOnNegI32);
 
   JSContext* cx = instance->cx();
   Table& table = *instance->tables()[tableIndex];
@@ -1062,26 +1056,12 @@
     return -1;
   }
 
-  switch (table.repr()) {
-    case TableRepr::Ref:
-      table.fillAnyRef(index, 1, AnyRef::fromCompiledCode(value));
-      break;
-    case TableRepr::Func:
-      MOZ_RELEASE_ASSERT(!table.isAsmJS());
-      table.fillFuncRef(index, 1, FuncRef::fromCompiledCode(value), cx);
-      break;
-  }
-
+  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
+  MOZ_RELEASE_ASSERT(!table.isAsmJS());
+  table.fillFuncRef(index, 1, FuncRef::fromCompiledCode(value), cx);
   return 0;
 }
 
-/* static */ uint32_t Instance::tableSize(Instance* instance,
-                                          uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableSize.failureMode == FailureMode::Infallible);
-  Table& table = *instance->tables()[tableIndex];
-  return table.length();
-}
-
 /* static */ void* Instance::refFunc(Instance* instance, uint32_t funcIndex) {
   MOZ_ASSERT(SASigRefFunc.failureMode == FailureMode::FailOnInvalidRef);
   JSContext* cx = instance->cx();
@@ -1427,7 +1407,7 @@
     const TableDesc& td = metadata().tables[i];
     TableInstanceData& table = tableInstanceData(td);
     table.length = tables_[i]->length();
-    table.functionBase = tables_[i]->functionBase();
+    table.elements = tables_[i]->tlsElements();
   }
 
 #ifdef ENABLE_WASM_EXCEPTIONS
@@ -2343,7 +2323,7 @@
     if (tables_[i] == theTable) {
       TableInstanceData& table = tableInstanceData(metadata().tables[i]);
       table.length = tables_[i]->length();
-      table.functionBase = tables_[i]->functionBase();
+      table.elements = tables_[i]->tlsElements();
     }
   }
 }